dataset/slice.cpp

```diff
@@ -0,0 +1,101 @@
+#include "scipp/dataset/slice.h"
+#include "scipp/dataset/dataset.h"
+#include "scipp/units/dim.h"
+#include "scipp/variable/comparison.h"
+#include "scipp/variable/reduction.h"
+#include "scipp/variable/util.h"
+#include "scipp/variable/variable.h"
+
+namespace scipp::dataset {
+DataArrayConstView slice(const DataArrayConstView &to_slice, const Dim dim,
+                         const VariableConstView begin,
+                         const VariableConstView end) {
+  using namespace variable;
+  if (begin)
+    variable::detail::expect0D(begin.dims());
+  if (end)
+    variable::detail::expect0D(end.dims());
+  auto coords = to_slice.coords();
+  if (!coords.contains(dim))
+    throw except::DimensionNotFoundError(to_slice.dims(), dim);
+  auto [k, coord] = *coords.find(dim);
+  if (coord.dims().ndim() != 1) {
+    throw except::SizeError(
+        "Multi-dimensional coordinates not supported in slice");
+  }
+  const bool ascending = is_sorted(coord, dim, variable::SortOrder::Ascending);
+  const bool descending =
+      is_sorted(coord, dim, variable::SortOrder::Descending);
+
+  if (!(ascending ^ descending))
+    throw std::runtime_error("Coordinate must be monotomically increasing or "
+                             "decreasing for value slicing");
+
+  const auto len_data = to_slice.dims()[dim];
+  const auto len = coord.dims().volume();
+  const auto bin_edges = len == len_data + 1;
+
+  // Point slice
+  if ((begin && end) && begin == end) {
```

Comment:
Not sure I understand the second part of the check, `begin == end` should return and empty range (dimension extent `0`), just like `data['x', 4:4]`.

Reply:
I see you are trying to support `begin` and `end` unspecified by passing an invalid view? I think that is fine, but then there is no way to distinguish from a point slice, so maybe we need a differently named method for that?